Mathlib/Algebra/Ring/Associator.lean

```diff
@@ -0,0 +1,77 @@
+/-
+Copyright (c) 2025 Bernhard Reinke. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bernhard Reinke
+-/
+import Mathlib.Algebra.Ring.Defs
+import Mathlib.Algebra.Group.Hom.End
+import Mathlib.Algebra.Group.Hom.Instances
+import Mathlib.Tactic.Abel
+
+/-!
+If `R` is a non-associative ring, then  `(x * y) * z - x * (y * z)` is called the `associator` of
+ring elements `x y z : R`. We realize the associator as a map that is an `AddMonoidHom` in each
+argument. The associator vanishes exactly when `R` is associative. We prove a similar statement for
+semirings by directly comparing the maps `mull₃` and `mulr₃`, the `AddMonoidHom` versions of the
+multiplications `(x * y) * z` and `x * (y * z)`.
+-/
+
+universe u
+
+namespace AddMonoidHom
+
+section NonUnitalNonAssocSemiring
+variable {R : Type u} [NonUnitalNonAssocSemiring R]
+
+/-- The multiplication `(x * y) * z` of three elements of a (non-associative)
+(semi)-ring is an `AddMonoidHom` in each argument. -/
+def mull₃ : R →+ R →+ R →+ R where
+  toFun x := comp mul (mulLeft x)
+  map_zero' := by ext; simp
+  map_add' := fun x y => by ext; simp [add_mul]
+
+@[simp]
+theorem mull₃_apply (x y z : R) : mull₃ x y z = (x * y) * z := by simp [mull₃]
+
+/-- The multiplication `x * (y * z)` of three elements of a (non-associative)
+(semi)-ring is an `AddMonoidHom` in each argument. -/
+def mulr₃ : R →+ R →+ R →+ R where
+  toFun x := compr₂ mul (mulLeft x)
+  map_zero' := by ext; simp
+  map_add' := fun x y => by ext; simp [add_mul]
+
+@[simp]
+theorem mulr₃_apply (x y z : R) : mulr₃ x y z = x * (y * z) := by simp [mulr₃]
+
+/-- An a priori non-associative semiring is associative if the `AddMonoidHom` versions of
+the multiplications `(x * y) * z` and `x * (y * z)` agree. -/
+theorem associative_iff_mull₃_eq_mulr₃ : (∀ (x y z : R), (x * y) * z = x * (y * z)) ↔
+  mull₃ (R := R) = mulr₃ := by
+  constructor
+  · intro h
+    ext x y z
+    simp [h x y z]
+  · intro h x y z
+    rw [← mull₃_apply, ← mulr₃_apply, h]
```

Comment:
I think it might be worth restating this in terms of `Std.Associative`, e.g something of the form
```suggestion 
theorem find_me_name : mull₃ (R := R) = mulr₃ ↔
    Std.Associative ((· * ·) : R → R → R) where
  mp := fun h ↦ ⟨fun x y z => by rw [← mulr₃_apply,← mull₃_apply, h]⟩
  mpr := fun h ↦ by ext ; apply h.assoc
```

Reply:
I think it might also be worth adding this implication separately 
```suggestion 
lemma mull₃_eq_mulr₃ [NonUnitalSemiring R] : mull₃ (R := R) = mulr₃ := by 
    sorry
```